components/notebook-controller/controllers/notebook_controller.go

```diff
@@ -567,7 +568,142 @@ func generateVirtualService(instance *v1beta1.Notebook) (*unstructured.Unstructu
+			"namespace": gatewayNamespace,
+		},
+	}
+	if err := unstructured.SetNestedSlice(httpRoute.Object, parentRefs, "spec", "parentRefs"); err != nil {
+		return nil, fmt.Errorf("set .spec.parentRefs error: %v", err)
+	}
+
+	// Set rules for HTTPRoute
+	rules := []interface{}{
+		map[string]interface{}{
+			"matches": []interface{}{
+				map[string]interface{}{
+					"path": map[string]interface{}{
+						"type":  "PathPrefix",
+						"value": prefix,
+					},
+				},
+			},
+			"filters": []interface{}{
+				map[string]interface{}{
+					"type": "URLRewrite",
+					"urlRewrite": map[string]interface{}{
+						"path": map[string]interface{}{
+							"type":               "ReplacePrefixMatch",
+							"replacePrefixMatch": pathRewrite,
+						},
+					},
+				},
+			},
+			"backendRefs": []interface{}{
+				map[string]interface{}{
+					"name": name,
+					"port": int64(DefaultServingPort),
+				},
+			},
+		},
+	}
+	if err := unstructured.SetNestedSlice(httpRoute.Object, rules, "spec", "rules"); err != nil {
+		return nil, fmt.Errorf("set .spec.rules error: %v", err)
+	}
 
+	return httpRoute, nil
+}
+
+func (r *NotebookReconciler) reconcileRouting(instance *v1beta1.Notebook) error {
+	useAmbientMode := os.Getenv("USE_AMBIENT_MODE") == "true"
+	
+	if useAmbientMode {
+		return r.reconcileHTTPRoute(instance)
+	} else {
+		return r.reconcileVirtualService(instance)
+	}
+}
+
+func (r *NotebookReconciler) reconcileHTTPRoute(instance *v1beta1.Notebook) error {
+	log := r.Log.WithValues("notebook", instance.Namespace)
+	httpRoute, err := generateHTTPRoute(instance)
+	if err != nil {
+		log.Info("Unable to generate HTTPRoute...", err)
+		return err
+	}
+	if err := ctrl.SetControllerReference(instance, httpRoute, r.Scheme); err != nil {
+		return err
+	}
+	// Check if the HTTPRoute already exists.
+	foundHTTPRoute := &unstructured.Unstructured{}
+	justCreated := false
+	foundHTTPRoute.SetAPIVersion("gateway.networking.k8s.io/v1")
+	foundHTTPRoute.SetKind("HTTPRoute")
+	err = r.Get(context.TODO(), types.NamespacedName{Name: virtualServiceName(instance.Name,
+		instance.Namespace), Namespace: instance.Namespace}, foundHTTPRoute)
+	if err != nil && apierrs.IsNotFound(err) {
+		log.Info("Creating HTTPRoute", "namespace", instance.Namespace, "name",
+			virtualServiceName(instance.Name, instance.Namespace))
+		err = r.Create(context.TODO(), httpRoute)
+		justCreated = true
+		if err != nil {
+			return err
+		}
+	} else if err != nil {
+		return err
+	}
+
+	if !justCreated {
+		// Copy the spec from the desired HTTPRoute to the existing one
+		if !reflect.DeepEqual(httpRoute.Object["spec"], foundHTTPRoute.Object["spec"]) {
+			log.Info("Updating HTTPRoute", "namespace", instance.Namespace, "name",
+				virtualServiceName(instance.Name, instance.Namespace))
+			foundHTTPRoute.Object["spec"] = httpRoute.Object["spec"]
+			err = r.Update(context.TODO(), foundHTTPRoute)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
 }
```

Comment:
Looks good, but I 'd like an approval from another reviewer too on this reconcile function.